server/src/main/java/org/apache/iotdb/db/query/udf/core/transformer/BinaryTransformer.java

```diff
@@ -0,0 +1,147 @@
+
+  private final LayerPointReader leftPointReader;
+  private final LayerPointReader rightPointReader;
+
+  protected BinaryTransformer(LayerPointReader leftPointReader, LayerPointReader rightPointReader) {
+    this.leftPointReader = leftPointReader;
+    this.rightPointReader = rightPointReader;
+  }
+
+  @Override
+  public boolean isConstantPointReader() {
+    return leftPointReader.isConstantPointReader() && rightPointReader.isConstantPointReader();
+  }
+
+  @Override
+  protected boolean cacheValue() throws QueryProcessException, IOException {
+    if (!leftPointReader.next() || !rightPointReader.next()) {
+      return false;
+    }
+    if (!cacheTime()) {
+      return false;
+    }
+    if (leftPointReader.isCurrentNull() || rightPointReader.isCurrentNull()) {
+      currentNull = true;
+    } else {
+      switch (getDataType()) {
+        case DOUBLE:
+          cachedDouble =
+              evaluateDouble(
+                  castCurrentValueToDoubleOperand(leftPointReader),
+                  castCurrentValueToDoubleOperand(rightPointReader));
+          break;
+        case BOOLEAN:
+          cachedBoolean =
+              evaluateBoolean(
+                  castCurrentValueToDoubleOperand(leftPointReader),
+                  castCurrentValueToDoubleOperand(rightPointReader));
+      }
+    }
+    leftPointReader.readyForNext();
+    rightPointReader.readyForNext();
+    return true;
+  }
+
+  /**
+   * finds the smallest, unconsumed timestamp that exists in both {@code leftPointReader} and {@code
+   * rightPointReader} and then caches the timestamp in {@code cachedTime}.
+   *
+   * @return true if there has a timestamp that meets the requirements
+   */
+  private boolean cacheTime() throws IOException, QueryProcessException {
+    if (leftPointReader.isConstantPointReader() && rightPointReader.isConstantPointReader()) {
+      return true;
+    }
+    if (leftPointReader.isConstantPointReader()) {
+      cachedTime = rightPointReader.currentTime();
+      return true;
+    }
+    if (rightPointReader.isConstantPointReader()) {
+      cachedTime = leftPointReader.currentTime();
+      return true;
+    }
+
+    long leftTime = leftPointReader.currentTime();
+    long rightTime = rightPointReader.currentTime();
+
+    while (leftTime != rightTime) {
+      if (leftTime < rightTime) {
+        leftPointReader.readyForNext();
+        if (!leftPointReader.next()) {
+          return false;
+        }
+        leftTime = leftPointReader.currentTime();
+      } else {
+        rightPointReader.readyForNext();
+        if (!rightPointReader.next()) {
+          return false;
+        }
+        rightTime = rightPointReader.currentTime();
+      }
+    }
+
+    // leftTime == rightTime
+    cachedTime = leftTime;
+    return true;
+  }
+
+  protected double evaluateDouble(double leftOperand, double rightOperand) {
+    return 0.0;
+  }
+
+  protected boolean evaluateBoolean(double leftOperand, double rightOperand) {
```

Comment:
Could `evaluateBoolean` accept two boolean parameters instead of double?

Reply:
for compare expression like `a>b`, `a` `b` are both double. Only in logic `and` and `or`, rightOperand and rightOperand are boolean

Reply:
So is this defined in CompareBinaryTransformer better?